packages/middleware-sdk-s3/src/region-redirect-middleware.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  HandlerExecutionContext,
+  InitializeHandler,
+  InitializeHandlerArguments,
+  InitializeHandlerOptions,
+  InitializeHandlerOutput,
+  InitializeMiddleware,
+  MetadataBearer,
+  Pluggable,
+  Provider,
+  RelativeMiddlewareOptions,
+  SerializeHandler,
+  SerializeHandlerArguments,
+  SerializeHandlerOutput,
+  SerializeMiddleware,
+} from "@smithy/types";
+
+/**
+ * @internal
+ */
+interface PreviouslyResolved {
+  region: Provider<string>;
+  followRegionRedirects: boolean;
+}
+
+/**
+ * @internal
+ */
+export function regionRedirectMiddleware(clientConfig: PreviouslyResolved): InitializeMiddleware<any, any> {
+  return <Output extends MetadataBearer>(
+      next: InitializeHandler<any, Output>,
+      context: HandlerExecutionContext
+    ): InitializeHandler<any, Output> =>
+    async (args: InitializeHandlerArguments<any>): Promise<InitializeHandlerOutput<Output>> => {
+      try {
+        return next(args);
+      } catch (err) {
+        if (
+          clientConfig.followRegionRedirects &&
+          err.Code === "PermanentRedirect" &&
```

Comment:
Can we use `.name` here?
```suggestion
          err.name === "PermanentRedirect" &&
```

The `.Code` was used in v2, and we recommended using `.name` before introducing [Service Exceptions with classes](https://aws.amazon.com/blogs/developer/service-error-handling-modular-aws-sdk-js/).